Product management

FAEZ10 opened the pull request:
# **Pull Request: Product Management Enhancements**

## **Type of Change**
- [x] Feature
- [ ] Fix

---

## **Description**

This PR introduces multiple enhancements and new features for product and product-related functionalities, including CRUD operations, advanced methods, and service integrations. Below is a summary of the changes:

### **1. Product Management Enhancements**
- Implemented `toggleProductActivation` to activate or deactivate products.
- Added `searchProducts` for advanced filtering based on name, categories, and price range.
- Introduced `duplicateProduct` to create a copy of an existing product.
- Added `getProductsByStatus` to filter products by their activation status.
- Developed `addImagesToProduct` to associate images with products.
- Created `removeImageFromProduct` to dissociate and delete images from products.

### **2. Product Variant Management**
- Implemented CRUD operations for managing product variants.
- Added `searchVariants` for advanced filtering based on product, size, color, and price range.
- Introduced `getLimitedEditionVariants` to retrieve variants marked as limited editions.
- Created `duplicateVariant` to clone an existing product variant.

### **3. Product Features and Categories**
- Developed `ProductHasCategoryService`:
  - `linkProductToCategory` to associate a product with a category.
  - `unlinkProductFromCategory` to remove associations between products and categories.
- Developed `ProductHasProductFeatureService`:
  - `addFeatureToProduct` to associate features with products and assign values.
  - `removeFeatureFromProduct` to dissociate features from products.

### **4. Product Images**
- Added CRUD operations for managing product images.
- Integrated routes for creating, updating, retrieving, and deleting images.
- Enhanced validation to ensure consistency in image-product relationships.

---

## **How Has This Been Tested?**

- [x] Manually tested all API endpoints using Postman.
- [x] Verified relationships between products, categories, and features.
- [x] Ensured proper association and dissociation of images with products.
- [x] Tested all CRUD operations for Product and ProductVariant services.
- [x] Validated advanced search and filtering functionalities.

---

## **Checklist**
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation (Swagger).
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

---

## **Relevant Issues or Context**

This PR consolidates the development of features for the **`product-management`** branch, including advanced methods and CRUD operations for the following:
- Products
- Product Variants
- Product Features
- Product Categories
- Product Images

FAEZ14 committed:
Fix conflict

FAEZ14 committed:
feat(product-feature): Implement CRUD operations for ProductFeature

- Added routes for creating, reading, updating, and deleting product features.
- Integrated middleware for validation, authentication, and role-based access control.
- Included Swagger documentation for all API endpoints.
- Ensured DTO validation for create and update operations.
- Added support for retrieving all product features and individual features by ID.

FAEZ14 committed:
feat(product-image): Implement CRUD operations for ProductImage

- Added routes for creating, retrieving, updating, and deleting product images.
- Integrated middleware for validation, authentication, and role-based access control.
- Included Swagger documentation for all API endpoints.
- Ensured DTO validation for create and update operations.
- Added support for retrieving all images for a specific product and updating individual images by ID.

FAEZ14 committed:
feat(product): Add advanced features to ProductService

- Implemented `toggleProductActivation` to enable or disable a product.
- Added `searchProducts` for advanced product search based on name, categories, and price range.
- Implemented `duplicateProduct` to create a copy of an existing product.
- Added `getProductsByStatus` to filter products by their activation status.
- Introduced `addImagesToProduct` for associating multiple images to a product.
- Added `removeImageFromProduct` to dissociate and delete images from a product.
- Improved validation logic for product dependencies (categories, features, images, and user).

FAEZ14 committed:
feat(product-variant): Implement CRUD operations and additional features for ProductVariant

- Added CRUD operations for managing product variants:
  - Create, read (single and all), update, and delete product variants.
  - Integrated relation handling with products, sizes, and colors.
- Implemented `searchVariants` for advanced filtering based on product, size, color, and price range.
- Added `getLimitedEditionVariants` to retrieve product variants marked as limited edition.
- Introduced `duplicateVariant` to create a copy of an existing product variant.
- Enhanced dependency validation for related entities (products, sizes, colors).
- Improved data retrieval by populating related entities in all relevant methods.

FAEZ14 committed:
feat(product-category): Implement service for managing product-category relationships

- Added `linkProductToCategory` to associate a product with a category.
- Added `unlinkProductFromCategory` to remove an association between a product and a category.
- Introduced validation to ensure `productId` and `categoryId` are valid ObjectIds.
- Integrated a repository layer for database operations.

FAEZ14 committed:
feat(product-feature): Implement service for managing product-feature relationships

- Added `addFeatureToProduct` to associate a feature with a product and assign a value.
- Added `removeFeatureFromProduct` to delete an association between a product and a feature.
- Introduced validation to ensure `productId` and `featureId` are valid ObjectIds.
- Integrated a repository layer for database operations.
- Improved error handling for invalid ObjectIds.

FAEZ14 committed:
Fix getImagesByProductId route

FAEZ14 committed:
fix(product, product-variant): Resolve route collision and standardize field naming

- Fixed the route order issue in Product routes to avoid conflicts between dynamic routes (`/:id`) and specific routes (`/search`).
- Renamed `product_id`, `size_id`, and `color_id` to `productId`, `sizeId`, and `colorId` for consistency across ProductVariant models and services.
- Updated repository methods in ProductVariantRepository to use the standardized field names:
  - `getAllWithRelations`
  - `findByIdWithRelations`
  - `findByCriteria`
  - `findLimitedEditionVariants`
- Improved route handling in Product and ProductVariant modules to align with updated naming conventions.

johnnyhelloworld reviewed (approved):
Review code + Testes des routes sur Postman OK

DaniOmer commented on src/apps/product/api/product.route.ts:
```diff
@@ -69,7 +73,45 @@ router.post(
+ *       200:
+ *         description: Products retrieved successfully
+ */
+router.get("/search", authenticateMiddleware, productController.searchProducts.bind(productController));
```
Je pense qu'un user non connecté devrait pouvoir rechercher un produit

DaniOmer commented on src/apps/product/api/product.route.ts:
```diff
@@ -96,6 +138,7 @@ router.get("/", productController.getAllProducts.bind(productController));
+  authenticateMiddleware,
   validateIdMiddleware("Product"),
   productController.getProductById.bind(productController)
 );
```
Pareil ici, normal on peut afficher un produit sans être connecté

DaniOmer commented on src/apps/product/api/productFeature/productFeature.route.ts:
```diff
@@ -0,0 +1,57 @@
+
+router.get("/",
+    authenticateMiddleware,
+   productFeatureController.getAll.bind(productFeatureController));
```
Idem

DaniOmer commented on src/apps/product/data-access/product.model.ts:
```diff
@@ -1,10 +1,20 @@
-export const ProductModel = model<IProduct>("Product", ProductSchema);
\ No newline at end of file
+    description: { type: String },
+    basePrice: { type: Number },
```
Ici il faut qu'on ait la même chose que dans product variant `priceEtx` et `priceVat`

DaniOmer commented on src/apps/product/domain/product.service.ts:
```diff
@@ -35,4 +69,169 @@ export class ProductService extends BaseService {
+      ...product.toObject(),
+      _id: undefined,
+      name: `${product.name} (Copy)`,
+    };
```
je ne sais pas si c'est pertinent le duplicate product. Mais niveau code pour faire plus propre tu devrais destructurer autrement, récupérer les champs que tu peux de product sans prendre id, _id, createdAt et updatedAt

DaniOmer commented on src/apps/product/domain/productVariant/productVariant.service.ts:
```diff
@@ -34,4 +47,87 @@ export class ProductVariantService extends BaseService {
+      ...variant?.toObject(),
+      _id: undefined,
+      stripeId: undefined, 
+    };
```
idem que pour product

FAEZ14 committed:
feat(tests): setup comprehensive configuration for unit and integration tests

- Added configuration for unit tests with Jest.
- Implemented integration test setup with database cleanup and seeding utilities.
- Configured test scripts in `package.json`:
  - `test` for running all tests.
  - `test:unit` for unit tests.
  - `test:integration` for integration tests.
- Updated database utilities for test isolation:
  - Automatic cleanup after each test.
  - Seeding of initial data for integration tests.
- Set up a dedicated MongoDB service for testing in `docker-compose.yml`:
  - `mongo_test` service on port `27018` for isolation.
  - Configured authentication with dynamic environment variables.
  - Enabled `--auth` mode for stricter security during tests.
- Enhanced reusability with mock data, helpers, and clear separation of test environments.

FAEZ14 committed:
feat(tests): add unit tests for auth-related modules

- Added unit tests for the following modules:
  - Role Service
  - User Service
  - Permission Service
  - Auth Service
  - Security Helper functions

- Ensured comprehensive test coverage for:
  - Role creation, retrieval, and validation
  - User authentication and profile management
  - Permission assignment and validation
  - Token generation and verification in the security helper
  - Error handling and edge cases for all modules

- Improved test setup and teardown processes to ensure clean database states.

FAEZ14 committed:
test(integration): add integration tests for auth-related modules

- Implemented integration tests for the following endpoints:
  - Role Routes (creation, retrieval)
  - User Routes (CRUD operations)
  - Permission Routes (creation, validation)
  - Auth Routes (registration, login)

- Ensured full integration coverage for:
  - API request/response validation
  - Authentication and authorization workflows
  - Role and permission management logic
  - Error handling and edge case scenarios

- Enhanced test database setup with proper seeding and cleanup for consistent results.
